### PR TITLE
fix: Fix hashing of user ip addresses

### PR DESCRIPTION
**Pull request by untitaker**
If the IP address has become invalid as part of PII processing, we move it into the user ID. That ensures people can do IP hashing and still have a correct users-affected count.

Right now both Snuba and EventUser discard unparseable IPs for indexing, and we assume we want to keep it that way.

If there are any other fields set that take priority over the IP for uniquely identifying a user (has_other_fields), we do not want to do anything. The value will be wiped out in renormalization anyway.

Follow-up to https://github.com/getsentry/relay/pull/686

### CODE REVIEW

File: relay-general/src/pii/processor.rs [hunk 1]
```diff
@@ -139,6 +140,41 @@ impl<'a> Processor for PiiProcessor<'a> {
...
+        // Right now both Snuba and EventUser discard unparseable IPs for indexing, and we assume
+        // we want to keep it that way.
+        //
+        // If there are any other fields set that take priority over the IP for uniquely
+        // identifying a user (has_other_fields), we do not want to do anything. The value will be
+        // wiped out in renormalization anyway.
+        if ip_was_valid && !has_other_fields && !ip_is_still_valid {
+            let mut user_ip = Annotated::empty();
+            mem::swap(&mut user.ip_address, &mut user_ip);
+            user.id = user_ip.map_value(IpAddr::into_inner).map_value(From::from);
```
Comment (jan-auer) [inline relay-general/src/pii/processor.rs]: You can simplify this with:

```suggestion
            user.id = mem::take(&mut user.ip_address)
                .map_value(|ip| ip.into_inner().into());
```

File: relay-general/src/pii/processor.rs [hunk 2]
```diff
@@ -879,3 +915,38 @@ fn test_logentry_value_types() {
...
+
+    let user = event.value().unwrap().user.value().unwrap();
+
+    assert!(user.ip_address.value().is_none());
+
+    assert_eq!(
+        user.id.value().unwrap().as_str(),
+        "AE12FE3B5F129B5CC4CDD2B136B7B7947C4D2741"
+    );
+}
```
Comment (jan-auer) [inline relay-general/src/pii/processor.rs:950]: Could you please add a second test for the other case that this PR is supposed to solve: A user with an explicit ID that does not get overwritten.

**Review by jan-auer — APPROVED** [review]
Thank you, LGTM. One request for tests and a minor suggestion attached.

File: relay-general/src/pii/processor.rs (hunk 2 above)
Comment (untitaker) [reply relay-general/src/pii/processor.rs:950]: done

**Pull request merged**